src/Microsoft.TemplateEngine.Abstractions/Json/IJsonObject.cs

```diff
@@ -11,6 +11,10 @@ public interface IJsonObject : IJsonToken
 
         IJsonObject SetValue(string propertyName, IJsonToken value);
 
-        IReadOnlyCollection<string> ExtractValues(params (string propertyName, Action<IJsonToken> valueExtractor)[] mappings);
+        ISet<string> ExtractValues(IReadOnlyDictionary<string, Action<IJsonToken>> mappings);
+
+        ISet<string> ExtractValues<T>(T context, IReadOnlyDictionary<string, Action<IJsonToken, T>> mappings);
+
+        IEnumerable<KeyValuePair<string, IJsonToken>> Properties();
```

Comment:
Is there a reason to not have this return an IReadOnlyDictionary<string, IJsonToken> ?
It seems that the keys are necessarily unique.